<h1 style="display:none;">

# Objects, Variables, Types, Operations I


## First Program (Again)

### Overview

First simple program
    - Input the radius of a circle
    - Compute the
        - Circumfrence
        - Diameter
        - Area
        
### Implementation


In [16]:

"""
Created on Mon Jan  1 10:38:37 2018

@author: donaldferguson
"""

# This program allows a user to input the radius on a circle.
# We want to teach the formula to young children. So, we only
# allow the radius to be an integer.

# Almost every program you write will use "programs" others have written.
# Your successful programs will become programs that others use.
# Any non-trivial program requires a team. The team members assemble
# the solution from individual subcomponents they build.
# The subcomponents and reusable parts are called modules.

import math     # We just imported our first module.

# Programs, like mathematical functions, are only useful if they
# operate on many user provided inputs. To start, we will get the input from
# the "command line."

# Print a prompt asking for the radius.
# Set a variable to the input value.
radius_str = input('Enter the radius of a circle: ')

# Let's double check that we got the input.
print("You entered ", radius_str, " which is of type ", type(radius_str))

# We are going to do 'math' on the input. So, we should
# covert it to an Integer.
radius_int = int(radius_str)

# The circumfrence is 2 times pi time the radius.
# The area is pi * r squared.
circumference = 2 * math.pi * radius_int
area = math.pi * (radius_int ** 2)

# Python conventions do not like lines that are too long.
# \ means that we will continue the command on the next line.
print ("The cirumference is:",circumference,  \
      ", and the area is:",area)

Enter the radius of a circle: 3
You entered  3  which is of type  <class 'str'>
The cirumference is: 18.84955592153876 , and the area is: 28.274333882308138


## What Just Happened?

- We imported a pre-built module (math)
    - We are engineers and applied scientists after all.
    - Amateurs argue about the relative benefits of programming languages.
    - _A "language's" primary benefits are the libraries that simplify programming._


- A _program_ produces a computational result
    - (input) -> program -> (outcome)
    - The simplest input is from the "command line," aka "the terminal."
    - The simplest output is also the command line or terminal.
    - Most of the time the input comes from a graphical user interface form, dataset, etc.
    - Most of the time the output goes to a graphical user interface table, plot/diagram, dataset, etc.
    
    
- And, we used __variables__ of a couple of __types__
    
    

## Variables, Memory, Code ...

### Variables

<img src="../images/python-variables.png" width="90%">




### How a Computer Works (I)

<img src="../images/cpu1.jpg">
(http://csbapp.uncw.edu/mis213/TG01/TG1-5-A.html)

- The brain of a computer is the central processing unit (CPU) and internal, random access memory (RAM). RAM contains
    - Data
    - Instructions/commands
    - Everything is in binary (base 2 numbers).
<br><br>

<img src="../images/cpu2.jpg">
(http://csbapp.uncw.edu/mis213/TG01/TG1-5-A.html)

- The basic behavior of the CPU is
    1. Load the next instruction.
    2. Access two memory locations.
    3. Perform an operation using the bits in the memory locations.
    4. Store the result in a 3rd memory location.
    5. Go back to step 1.


- Steps 1-5 occur every "clock tick." 
    - My MacBook's clock speed is 2.7 GHz (gigahertz).
    - The CPU executes steps (1-5) 2.7 billion times per second.
    
<br><br>
A computer's RAM looks something like this ...

<img src="../images/memory2.jpeg">

    


- All data storage is in _binary_
    - The _number_ 217 is 0000 0000 0000 0000 0000 0000 1101 1001
    - This is awkward. So, we use hexadecimal digits 0,1,2,3,4,5,6,7,8,9,A,B,C,D,E,F
    - And 217 would be x000000D9
    - But, the _string_ "217" is x323137 and
    - The prompt for input "Enter the radius of a circle: " is 456e7465722074686520726164697573206f66206120636972636c653a20
    
    
- Instructions and commands are also in binary.
<br><br>
<img src="../images/x86instruction.png">


<br><br>
- This can get kind of tedious. People actually used to program computers this way, mechanically ...

<img src="../images/eniac_wires.jpg">

<br><br>
- And this is where the word "bug" comes from
<br>
<img src="../images/firstbug.jpg">


But, ...
- There are special programs called _compilers_ or _interpreters_ that
- Take input strings like
    - a = 3
    - b = a
    - c = b + a
- And convert into the much more tedious binary operations.

<img src="../images/compiler.png">

## Python Variable Names

- Variable name rules
    - Must begin with a letter (a - z, A - B) or underscore (_)
    - Other characters can be letters, numbers or _
    - Case Sensitive (Cat, CAt, cAt, ... are all different names)
    - Can be any (reasonable) length
    - There are some reserved words which you cannot use as a variable name because Python uses them for other things.
    
    
- Programs should be readable, at least to other programmers, which means that variable names should be intuitive.
    - There are programs executing right now.
    - That need to be modified by programmers.
    - And the original programmers retired or died a long time ago.
    - So,
        - zp_3R = zp_3R + (1 + qm97)* zp_3R is not good.
        - balance = balance + balance * (1 + interest_rate) is better.
        
        
- The variable name convention is "Lowercase with words separated by underscores as necessary to improve readability."


- There are recommended style guidelines for writing code, e.g.
    - https://google.github.io/styleguide/pyguide.html#Naming
    - https://www.python.org/dev/peps/pep-0008/
    - __The quality and style of your programs will define your reputation as a programmer.__  

## Let's Continue the Example

### The Example Program


In [17]:
# This program allows a user to input the radius on a circle.
# We want to teach the formula to young children. So, we only
# allow the radius to be an integer.

# Almost every program you write will use "programs" others have written.
# Your successful programs will become programs that others use.
# Any non-trivial program requires a team. The team members assemble
# the solution from individual subcomponents they build.
# The subcomponents and reusable parts are called modules.

import math     # We just imported our first module.

# Programs, like mathematical functions, are only useful if they
# operate on many user provided inputs. To start, we will get the input from
# the "command line."

# Print a prompt asking for the radius.
# Set a variable to the input value.
radius_str = input('Enter the radius of a circle: ')

# Let's double check that we got the input.
print("You entered ", radius_str, " which is of type ", type(radius_str))

# We are going to do 'math' on the input. So, we should
# covert it to an Integer.
radius_int = int(radius_str)

# The circumfrence is 2 times pi time the radius.
# The area is pi * r squared.
circumference = 2 * math.pi * radius_int
area = math.pi * (radius_int ** 2)

# Python conventions do not like lines that are too long.
# \ means that we will continue the command on the next line.
print ("The cirumference is:",circumference,  \
      ", and the area is:",area)



Enter the radius of a circle: 3
You entered  3  which is of type  <class 'str'>
The cirumference is: 18.84955592153876 , and the area is: 28.274333882308138


### Discussion

- Python variables reference (address, point to) "objects."
    - Unlike Java and many other languages, _variables do not have (declared) types._
    - The referenced objects do have types.
    - _Logically, the current type of the variable is the type of the object it references._
    
    
- This simple program uses 3 object types. The variables that reference the objects are:
    - Integer (radius_int)
    - Floating Point (circumference, area, math.pi)
    - String (radius_str)
    
## Variables, Names and Objects

### Example 1
    

In [18]:
my_variable = "Hello" 
print("my_variables is currently = ", my_variable, " of type ", type(my_variable))
my_variable = 2
print("my_variables is currently = ", my_variable, " of type ", type(my_variable))
my_variable = 3.1416
print("my_variables is currently = ", my_variable, " of type ", type(my_variable))


my_variables is currently =  Hello  of type  <class 'str'>
my_variables is currently =  2  of type  <class 'int'>
my_variables is currently =  3.1416  of type  <class 'float'>


### Example 2: Similar, more complex example

In [19]:
# "Pay no attention to the code behind that curtain, or def
def what(obj, callingLocals=locals()):
    for k,v in list(callingLocals.items()):
        if v is obj:
            name = k
            print("The object's name is ", name)
            print("The object's value is ", v)
            print("The object's ID is ", id(obj))
            print("The object's type is ", type(obj))
    

my_int1 = 10
my_int2 = 10
my_int3 = 8
my_int3 = my_int3 + 2

# I am going to print out some information about my_int1
print("\nTesting for my_int1")
what(my_int1)



Testing for my_int1
The object's name is  my_int1
The object's value is  10
The object's ID is  4336138576
The object's type is  <class 'int'>
The object's name is  my_int2
The object's value is  10
The object's ID is  4336138576
The object's type is  <class 'int'>
The object's name is  my_int3
The object's value is  10
The object's ID is  4336138576
The object's type is  <class 'int'>


What happened?
- That is not what I expected. What happened?
- We imprecisely use the term "variable" for things like my_int1.
- my_int1 is a "name." my_int1 = 10 means, "The object (thing) 10 is also known as my_int1."
- The statements my_int2 = 10 means, "The object (thing) 10 is also known as my_int2"
- The statements my_int3 = 8, my_int3 = my_int3 + 2 mean, "Take the name my_int3 away from 8 and assign to 10."
- There is a single "thing" (object, place in memory) that holds the value 10 but has three names.

### Example 3


In [20]:
my_int3 = "10"
my_int4 = "10"

# I am going to print out some information about my_int1
print("\nTesting for my_int3")
what(my_int3)



Testing for my_int3
The object's name is  my_int3
The object's value is  10
The object's ID is  4342734048
The object's type is  <class 'str'>
The object's name is  my_int4
The object's value is  10
The object's ID is  4342734048
The object's type is  <class 'str'>


OK. That seemed reasonable. The string "10" acted like the number 10, although using my_int3 for a string is poor.

### Example 4: Things get weird


In [21]:
my_int5 = 10
my_int6 = 10

my_int5 = str(my_int5)
my_int6 = str(my_int6)

print("my_int5 = ", my_int5, " of type ", type(my_int5))
print("my_int6 = ", my_int6, " of type ", type(my_int6))

# I am going to print out some information about my_int1
print("\nTesting for my_int5")
what(my_int5)

my_int5 =  10  of type  <class 'str'>
my_int6 =  10  of type  <class 'str'>

Testing for my_int5
The object's name is  my_int5
The object's value is  10
The object's ID is  4497771632
The object's type is  <class 'str'>


Huh. That test behaved significantly differently. I did not set my_int5 and my_int6 to "10". I computed two strings. This behavior is different from the previous examples.
- The statements
    - my_int3 = "10" and my_int4 = "10" mean
    - There is a specific, literal object "10".
    - my_int3 and my_int4 are additional names for that specific object "10"
- The statements
    - my_int5 = 10 and my_int6 = 10 mean
    - There is a specific, literal object 10.
    - my_int5 and my_int6 are additional names for that specific object 10
- The statements
    - my_int5 = str(my_int5) means:
        1. Find the thing with name my_int5.
        1. Make a new object of type string that has the string representation of the object named my_int5
    - my_int6 = str(my_int6) does the same for my_int6
- The result is that there are two __different__ string objects that happen to have the value "10", or you can think<br>
'There are two slots in memory, each with its own name, that just happen to hold the string "10"'
- This is a little confusing for basic types like String, Integer, etc. but is very important when we start dealing with more complex types.
        

### Summary

<img src="../images/variablesummary.jpeg" width="90%">
<br><br>

- Python is an object system and everything is an object.


- An object takes up space in memory and (may have) has data values in the space.<br>
In a computer, data and code are always represented in binary.


- The _type_ of the object defines how the Python runtime interprets/manages the bits. For example,
    - Decimal 21 is 0x00000015 = 00000000 00000000 00000000 00010101
    - Decimal 3.14 is 0x4048F5C3 = 01000000 01001000 11110101 11000011
    - "Veni, Vidi, Vici" is 
        - 0x56656E692C20566964692C2056696369 in hexadecimal
        - Binary
        
01010110011001010110111001101001001011000010000001010110011010010110010001101001001011000010000001010110011010010110001101101001


- Objects have types; names/variables do not have types.


- Many names can refer to the same object. A specific name
    - Refers to one object at a time.
    - Assignment statements, e.g. x = y + 1, change the object that the name references.
    
    
- Two different objects may have the same "value," i.e. bits and bytes.


- Every object has a unique, Python generated ID.

In [22]:
s1 = input("Caesar said? ")
s2 = input("Dixerunt Caesar?")

print("s1 = ", s1)
print("s2 = ", s2)

print("\n Examining s1 yields")
what(s1)
print("\n Examining s2 yields")
what(s2)

print("\n")
if (s1 == s2):
    print("s1 and s2 have the same value.")
if (id(s1) == id(s2)):
    print("s1 and s2 are the same object.")
else:
    print("s1 and s2 are NOT the same object.")

Caesar said? Veni, vidi, vici.
Dixerunt Caesar?Veni, vidi, vici.
s1 =  Veni, vidi, vici.
s2 =  Veni, vidi, vici.

 Examining s1 yields
The object's name is  s1
The object's value is  Veni, vidi, vici.
The object's ID is  4379353640
The object's type is  <class 'str'>

 Examining s2 yields
The object's name is  s2
The object's value is  Veni, vidi, vici.
The object's ID is  4379354144
The object's type is  <class 'str'>


s1 and s2 have the same value.
s1 and s2 are NOT the same object.


## Python Built-In Types

### Overview

<img src="../images/ptypes.jpeg">

<br><br>


### Numeric Types

- Numeric types are straight forward and are what you expect.


- There are a bunch of operators and functions, which we will cover later.


- There is a very powerful built-in math library, and some even more powerful optional packages.


- Some examples:

In [23]:
import math

x = 4
y = 3.1416
z = complex(x,y)
c = 1+2j

print("Some basic objects:")
print("x = ", x)
print("y = ", y)
print("z = ", z)
print("c = ", c)
print("\nResult of operations on objects.")
print("z + c \t = ", z+c)
print("x + y \t = ", x + y)
print("x * c \t = ", x*c)
print("\nResult of some more operations on objects.")
print("The real part of z is ", z.real)
print("The imaginary part of z  is ", z.imag)
print("The integer part of y is ", int(y))
print("I can also compute the ceiling of y, which is ", math.ceil(y))
print("x / y = ", x / y)
print("x // y =", x //y)




Some basic objects:
x =  4
y =  3.1416
z =  (4+3.1416j)
c =  (1+2j)

Result of operations on objects.
z + c 	 =  (5+5.1416j)
x + y 	 =  7.1416
x * c 	 =  (4+8j)

Result of some more operations on objects.
The real part of z is  4.0
The imaginary part of z  is  3.1416
The integer part of y is  3
I can also compute the ceiling of y, which is  4
x / y =  1.2732365673542145
x // y = 1.0



- Representing numberic types
    - int/Integer has 4 representations in Python
        - Decimal: 
            - Digits are 0,1,2,3,4,5,6,7,8,9
            - Examples are: 0, 231, 3978
        - [Hexadecimal](https://en.wikipedia.org/wiki/Hexadecimal):
            - Digits are 0,1,2,3,4,5,6,7,8,9,A,B,C,D,E,F
            - Examples:
                - 0x23 =  (2 * 16) + 3 = 35
                - 0xFA2 = (15 * (16 * 16) + (10 * 16) + 2  = 4002
        - [Octal](https://en.wikipedia.org/wiki/Octal)
            - Digits are 0,1,2,3,4,5,6,7
            - Examples:
                - 0o21 = (2 * 8) + 1 = 17
                - 0o1001 = 513
        - [Binary](https://en.wikipedia.org/wiki/Binary_number):
            - Digits are 0,1
            - Examples:
                - 0b11 = (1 * 2) + 1 = 3
                -0b1000 = 8
    - A real number has two representations
        - [Floating point decimal](https://en.wikipedia.org/wiki/Decimal_floating_point):
            - Format is (decimal int).(decimal int)
            - Examples:
                - 3.1416
                - 2.7813
                - 1.1414
        - [Exponential notation](https://en.wikipedia.org/wiki/Scientific_notation): Also known as scientific notation
            - Format is (floating point)e(int)
            - Examples:
                -  21e3 = 21000
                - 0.000213e3 = 0.213
        - [Complex](https://en.wikipedia.org/wiki/Complex_number):
            - Format is (floating point)+(floating point)j
            - Examples:
                -  2.1718+3.1416j = (2.1718+3.1416j)
                - complex(3,21.3) = (3+21.3j)
                
- Limits
    - Numbers in the real world can be -infinity to infinity, and have infinite number of decimal places.
    - The basic types (classes) in Python have upper and lower bounds.
    - There are modules that support increased limits and precision.
    
<img src="../images/floatinfo.jpeg">

- Rounding errors
    - In base 10, there is no decimal representation for irrational numbers, e.g. 1/3 or sqrt(2).
    - Some numbers are irrational in base 2, which is what a computer uses, e.g. 0.1 <br>So, (0.1 + 0.1 + 0.1) = 0.30000000000000004, and does not equal 0.3
    - This can break some fundamental laws of mathematics (algebra), e.g.
        - math.sqrt(2) * math.sqrt(2) = 2.0000000000000004 and not 2.
        -  0.3 - 0.1 - 0.1 - 0.1 = -2.7755575615628914e-17, not 0
        - (a + b) + c may or may not equal a + (b + c)
    - This is a behavior of ALL computers and programming languages.
    - In Python, you can handle with round()
        - 0.1 + 0.1 + 0.1 does not equal 0.3
        - round(0.1 + 0.1 + 0.1) DOES equal 0.3
        

In [24]:
x = 10e30
print("x = ", x)
y = -10e30
print("y = ", y)
print ("x + y = ", x + y)
print("\nNow it gets weird.")
print("(x + y) + 1 = ", (x + y) + 1)
print("x + (y + 1) = ", x + (y + 1))
print("\nYou can fix some of this with round()")
print("0.1 + 0.1 + 0.1 = ", 0.1 + 0.1 + 0.1)
print("round(0.1 + 0.1 + 0.1, 8) = ", round(0.1 + 0.1 + 0.1, 8))
print("math.sqrt(2) * math.sqrt(2) = ", math.sqrt(2) * math.sqrt(2))
print("round(math.sqrt(2) * math.sqrt(2), 10) = ", round(math.sqrt(2) * math.sqrt(2)))

x =  1e+31
y =  -1e+31
x + y =  0.0

Now it gets weird.
(x + y) + 1 =  1.0
x + (y + 1) =  0.0

You can fix some of this with round()
0.1 + 0.1 + 0.1 =  0.30000000000000004
round(0.1 + 0.1 + 0.1, 8) =  0.3
math.sqrt(2) * math.sqrt(2) =  2.0000000000000004
round(math.sqrt(2) * math.sqrt(2), 10) =  2


### Sequency Types

#### Overview and Text Sequences

<img src="../images/textsequence.jpeg">

The object type _str_, or text sequence, is an example of the basic sequence pattern
- There is an object of type "sequence." 


- In the example, the name of the object is _t_


- The object _t_ is itself made up of an iterable sequence of objects. The name of each object is logically  _t[i]_, e.g.
    - t[0]
    - t[1]
    - t[2]
    - ... ...
    

- A text sequence/str is simply a sequence in which each t[i] is a "text character."


- Remember,
    - In a computer, everything is a binary number.
    - For text and characters, the _encoding_ determines the mapping from integer value to printable character value.
    - [ASCII](https://en.wikipedia.org/wiki/ASCII) (American Standard Code for Information Interchange) used to be very common, but was "American" centric.
        - An ASCII only encodes 127 characters, and comes from the teletype days.
        - ASCII does support non-English characters.
    - IBM and some other systems used [Extended Binary Coded Decimal Interchange Code (EBCIDIC)](https://en.wikipedia.org/wiki/EBCDIC) for a long time.  
    - [UTF-8](https://en.wikipedia.org/wiki/UTF-8) is now the most common encoding.
        - UTF-8 encodes 1,112,064 characters from [Unicode](https://en.wikipedia.org/wiki/Unicode).
        - The first 127 numbers of ASCII and UTF-8 represent the same characters, that is UTF-8 is a superset of ASCII.
    
<img src="../images/encodinggrowth.jpeg">

_Some initial fun with encodings_
<br><br>

In [25]:
print("UTF (unicode) character number 20516 is = ", chr(20516))
print("There is no way to do this because ASCII is 1 byte and is 0 to 255")
print("ASCII is perfectly OK with 85 representing ", chr(85))

UTF (unicode) character number 20516 is =  値
There is no way to do this because ASCII is 1 byte and is 0 to 255
ASCII is perfectly OK with 85 representing  U


<br>
_Some function with text sequences (strings)_
<br>

In [26]:
mm = "A simple string"

# An sequence of bytes, which must be 0 .. 255
bb = bytes([87, 9, 88, 32, 89])
# A sequence of integers.
tt = [87, 9, 88, 32, 89, 20516, 1023]

print("\n")
print("mm = ", mm)
print("mm as a byte sequence = ", bytes(mm,"ascii"))
print("\n")
print("bb as a byte array = ", bb)
print('bb as a string = ', bb.decode("ascii"))
print('bb as a UTF string = ', bb.decode("utf8"))
print("\n")
print("Let's access parts of things.")
print("The fourth thing in mm is mm[3] which is ", mm[3])
print('bb as a string = ', bb.decode("ascii"), " and the fourth byte is ", bb[3])
print('bb as a string = ', bb.decode("ascii"), " and the second byte is ", bb[1])
print("bb[1] is the integer ", bb[1], " which is the tab character." )

print("\nNow some fun with sequence of integers")
print("tt = ", tt)
print("Let's try converting tt to a string. str(tt) = ", str(tt))
print("I got a string representation of the list.")
print("How about a string (text sequence) which converted each int to a UTF-8 chr?")
print("The answer is ''.join(map(chr,tt))", ''.join(map(chr,tt)), "\nand we will understand later.")



mm =  A simple string
mm as a byte sequence =  b'A simple string'


bb as a byte array =  b'W\tX Y'
bb as a string =  W	X Y
bb as a UTF string =  W	X Y


Let's access parts of things.
The fourth thing in mm is mm[3] which is  i
bb as a string =  W	X Y  and the fourth byte is  32
bb as a string =  W	X Y  and the second byte is  9
bb[1] is the integer  9  which is the tab character.

Now some fun with sequence of integers
tt =  [87, 9, 88, 32, 89, 20516, 1023]
Let's try converting tt to a string. str(tt) =  [87, 9, 88, 32, 89, 20516, 1023]
I got a string representation of the list.
How about a string (text sequence) which converted each int to a UTF-8 chr?
The answer is ''.join(map(chr,tt)) W	X Y値Ͽ 
and we will understand later.


#### Mutable and Immutable

"In object-oriented and functional programming, an _immutable object_ (unchangeable object) is an object whose state cannot be modified after it is created. This is in contrast to a _mutable object_ (changeable object), which can be modified after it is created." (https://en.wikipedia.org/wiki/Immutable_object)

- Most simple objects are immutable
    - You cannot change the integer 10 to 11, 12.37 or "Hello."
    - In the same way you cannot change 217345 to 341167, you cannot change "Hello" to "Halo."


- You can
    - Change what instance names reference.
    - Make new instances using parts of immutable objects.
    - etc.
    

- Mutability
    - Is an important concept, and its importance will become clear in future lectures.
    - We will not focus on the concept in the early lectures or projects, but you may occasionally hit find something that you think should work, but does not.



In [27]:
s = "Hello"
print ("s = ", s)

s2 = s[0] + "a" + s[3:5]
print("s2 = ", s2)

s =  Hello
s2 =  Halo


In [28]:
s[1]='a'

TypeError: 'str' object does not support item assignment

#### byte, bytearray, memoryview

- _Byte_ is the most basic unit of computer memory.
    - A byte is 8 bits (0/1 digits)
    - Decimal 0 to 255.
    - Binary 00000000 to 11111111
    - Hexadecimal 00,01, ... 09,0A, ..., OF, 10, 11, ..., 1F, ... FF.
        - A hex digit is 0 to F, which is 16 --> 4 bits.
        - A byte is 8 bits, or two hex digits.
        
        
- byte is immutable. byte_array _is_ mutable


- We will worry about _memoryview_ later


In [ ]:
x = b'El ni\xc3\xb1o come camar\xc3\xb3n'
print("x = ",x)
print("\nx.decode('utf-8)') = ",x.decode("utf-8"))
print("\nx to int big endian = ", int.from_bytes(x,  byteorder='big'))
print("\nx to int big endian = ", int.from_bytes(x,  byteorder='little'))

y = bytearray(x)
print("\ny = bytearray(c) produces y =", y)
print("\nn ASCII is the integer ", ord('n'))
y[5]=ord('n')
y[6]=ord('o')
y[7]=ord(' ')
print("I can change a bytearray. y[5]= the into ASCIii n, y[6] to ASCII o and y[7] to ASCII <space>, \nproduces y = ", y)
print("")

- 'enye' requires two bytes
    - I turned the first byte into 'n'
    - And the second by to < space >
- Why/when would you use _byte_, _bytearray_ and _memoryview?_
    - Integer, float, string, ... are by far the most common. Most of our projects will use these basic types and the other sequence types (below).
    - In many more advanced applications, you have to manipulate raw bytes, e.g.
        - Image, audio, etc. processing.
        - Cryptography
        - Signal processing
        
        

<img src="../images/jpeghex.jpeg">

#### Lists, Sets, Maps

"Lists are mutable sequences, typically used to store collections of homogeneous items (where the precise degree of similarity will vary by application)." (https://docs.python.org/3/library/stdtypes.html#list)

Many other languages use the term "array" for the language's similar concept.

__A simple example__


In [ ]:
x = [1, 10, 3.1416+2.7318j, "Hello", b'El ni\xc3\xb1o come camar\xc3\xb3n']
print("\nWhat is 'similar' about the elements of ", x, "?")
print("All are examples of built-in types used in the lecture")


__List are mutable__

In [ ]:
x = [1, 10, 3.1416+2.7318j, "Hello", b'El ni\xc3\xb1o come camar\xc3\xb3n']
print("\nx starts out as = ", x)
x[1]=bytes("Canary".encode("UTF-8"))
print("\nx becomes = ", x)

y = ["foo", 2e9]
print("\ny = ", y)

y = x + y
print("\ny becomes = ", y)
del(y[1])
print("\ny becomes = ", y)


__Lists are fundamental of data, analytics and plotting__

In [ ]:
import numpy as np 
import matplotlib.pyplot as plt


a = np.array([1,3]) 
print("\nVector a = ", a)
b = np.array([5,2])
print("Vector b = ", b)

print("\nThe inner product of a and b is = ", np.inner(a,b))

ad = np.add(a,b)
print("\nThe sum of a + b is the vector ", ad)

ax = plt.axes()
plt.axis([0, 7, 0, 7])
l1 = ax.arrow(0,0,a[0],a[1], color='r', label="a",head_width=0.2, head_length=0.2)
l2 = ax.arrow(0,0, b[0],b[1], color='b', label="b",head_width=0.2, head_length=0.2)
l3 = ax.arrow(0,0,ad[0],ad[1], color='g', label="a+b",head_width=0.2, head_length=0.2)
l4 = ax.arrow(a[0],a[1],(ad[0]-a[0]),(ad[1]-a[1]),head_width=0.2, head_length=0.2,color='c',label="b starting from  a")
legend = ax.legend(handles=[l1,l2, l3, l4],loc='upper left')

plt.show() 

### Homework 1, Part B

- Due date, points and full description are on [CourseWorks](https://courseworks2.columbia.edu/courses/53496/assignments/134478)


- Build a Jupyter notebook with the name "<uni>_assignment_2"
    

- Add the Python code from Assignment 1, Part A to a code cell in the Jupyter notebook.


- Build your UNI
    - Create a Code cell in the notebook. Use python statements like a=23 or b="foo".
    - Assign the names last_name, first_name, middle_name to the string values for your last, first and middle names.
    - The prefix of your UNI is some combination of the first character of the name string. For example, my full name is "Donald Francis Ferguson" and my UNI prefix is "dff".
    - The suffix of the UNI is a number. Assign the name uni_suffix to the integer number in your UNI. Mine would be the integer 9.
    - Using modifications from the code in lecture 1 and 2, create a string reprentation of your UNI and print it.
    - Uppercase letters are fine, e.g. "DFF9" is a correct submission.
    
- List
    - Create a list that contains your first name, middle name, last name and UNI.
    - Remove your middle name from the list.
    - Move the UNI to the first element in the list.
    
    